elkscmd/man/man5/dir.5
.TH DIR 5
.SH NAME
dir \- directory layout
.SH SYNOPSIS
.nf
.ft B
#include <dirent.h>
.SH DESCRIPTION
The directories of the MINIX and FAT file systems are arrays of the
following structure defined in <dirent.h>:
.PP
.nf
.ta +5n +15n +15n
#define MAXNAMLEN	26		   /* 14 for MINIX, 26 for FAT */
struct dirent {
	ino_t	d_ino;	        /* I-node number */
	unsigned short d_namlen;
	char	d_name[MAXNAMLEN+1];   /* File name */
};
.fi
.DT
.PP
The
.B d_ino
field is the inode number of the file named by
.BR d_name .
.B D_ino
is zero if the directory slot isn't allocated.  This number is the same as
.B st_ino
returned by
.BR stat (2)
unless the entry is mounted on.
.B D_name
is the name of up to 26 characters, null-terminated.
Any character other than null or '\fB/\fP' is allowed.
.PP
See
.BR directory (3)
for a portable way to access directories, ELKS and MINIX 3 are probably the last systems
with these old V7 format directories.
.SH "SEE ALSO"
.BR directory (3).
.SH AUTHOR
Kees J. Bot (kjb@cs.vu.nl)


elks/CHANGELOG
This file is used to record changes made to the CVS database. Please add to
the top of this file before committing any changes. Please use the header
format provided for each entry, noting that the text on each line should not
be longer than the lines of dashes.

Al Riddoch.

Contributor           Email address                     CHANGELOG entries found
~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~
Alan Cox              alan@redhat.com                   17 Jun 2001-18 Jun 2001
Alastair Bridgewater  nyef@softhome.net                 19 Nov 1998- 6 Dec 1998
Alistair Riddoch      alriddoch@zepler.org              25 Sep 1998-17 Jun 2001
Arnaud Launay         asl@launay.org                    15 Jul 1999-17 Oct 2001
Christian Mardmuller  chm@kdt.de                        10 Nov 1999-29 Feb 2000
Christoph Plattner    Christoph.Plattner@dot.at         31 Jan 2000- 2 Feb 2000
Claudio Matsuoka      claudio@conectiva.com
Daniele Giacomini     daniele@swlibero.org		20 Sep 2004
David Murn            scuffer@hups.apana.org.au         11 Nov 1998-15 Nov 1999
Eduardo Habkost	      ehabkost@conectiva.com.br		01 Jun 2004
Giles Russell                                           22 Nov 1999
Greg Haerr                                              14 Jul 1999-20 Jul 1999
Harry Kalogirou       HarKal@gmx.net                    20 Jun 2001- 2 Jan 2003
Jacek Lipkowski       sq5bpf@acid.ch.pw.edu.pl          20 Jan 2002
Jay Barr                                                27 Oct 2002
Ken Martwick          kenm@efn.org                      12 May 2002
M Warner Losh         imp@village.org                   20 Jan 2002
Manuel Novoa III      mjn3@codepoet.org                 18 Jun 2002-25 Jun 2002
Mario Frasca                                            20 Jan 2000-14 Feb 2000
Miguel Bolanos	      mike@linuxlabs.com		20 Sep 2004
Paul Nasrat           pauln@truemesh.com                29 Oct 2001-12 Jul 2002
Per Olofsson          MagerValp@cling.gu.se             25 Feb 2000
Peter L Vachuska      pvachusk@nconnect.net             13 Aug 2001
Richard Wallman       murkygoth@murkygoth.uklinux.net    2 Apr 2002
Riley Williams        Riley@Williams.Name               16 Nov 1998-13 Jul 2003
Roman Festchook       roma@polesye.net                  27 Oct 2001
Simon Weijgers        simon@weijgers.com                16 Sep 1998-15 Nov 1999
Simon Wood                                               6 Jan 2000-12 Jan 2000
Thomas McWilliams     tgm9999@yahoo.com                 27 Aug 1999- 2 Apr 2001
Tommy McCabe 	      rocketjet314@yahoo.com		31 May 2004

-------------------------------------------------------------------------------

Thu Feb 16 17:03:00 EST 2012	Jody Bruchon <jody@jodybruchon.com>	
	* Removed elksfs. It was a copy of the MINIX fs and no tools
	  exist to make such a filesystem.

Thu Nov 23 21:35:28 EST 2006	Segin <segin2005@gmail.com>
	* Changed name for the submenu for selecting charater
	  device drivers to note the fact that that submenu is broken;
	  attempting to enter it will have lxdialog die horribly.

Wed Nov  8 17:49:52 EST 2006	Segin <segin2005@gmail.com>

	* Changed ELKS init sequence to check all locations for 'init'
	  that the Linux kernel does. This is to follow more 
	  "traditional" behaviour.

Sun Aug 20 21:51:25 UTC 2006	Claudio Matsuoka <claudio@mandriva.com>

	* Allow loading Minix executables with total size larger than
	  64KB, such as ash (70140 bytes). Currently the total size
	  minus 32 bytes of header was limited to 16-bit.

	* Set tty1 back to minor 0. This is different from Linux
	  numbering, but it's necessary to have current ELKS virtual
	  terminals working. This and the executable size change put
	  ELKS back in usable state, allowing logins in the first VT
	  and executing a shell.

Mon Sep 20 23:27:23 CST 2004	Daniele Giacomini <daniele@swlibero.org>

	* Added italian keys support, Miguel Bolanos cleaned the original
	  file code.

Wed Jun  9 17:27:52 BRT 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* fix for my last fix: s/rp_cmd/rq_cmd/

Thu Jun 03 21:47:13 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* fixed format specifiers to correctly displaysegment sizes as unsigned
	  values, and to display debug messages correctly.

Thu Jun 03 07:48:25 CST 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* Fixed wrong struct field name for arch/i86/drivers/block/doshd.c
	* Several fixes ro arch/i86/mm/user.c

Web Jun 02 14:16:57 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* Added an extra fix to Eduardos memcpy fix.

Tue Jun 01 21:00:03 CST 2004	Eduardo Habkost ehabkost@conectiva.com.br

	* Fix memcpy_{to,from}fs, adding code to save the contents of SI and DI

Mon May 31 07:39:47 CST 2004	Tommy McCabe <rocketjet314@yahoo.com>

	* Fix wrong assignment (typo?) on printk
	* Fix order of LD parameters on arch/i86/Makefile
	* Use #ifdef rightly on setup.S
	* Fix init to use tty0, not tty1 as system console.

===============================================================================

Sun Jul 13 04:24:44 BST 2003	Riley Williams <Riley@Williams.Name>

	* This is ELKS 0.1.3-pre1

Sat Jul 12 12:24:36 BST 2003	Riley Williams <Riley@Williams.Name>

	* Rewrote kernel/printk.c to allow printk() to correctly print out
	  signed values.
	* Rewrote kernel/printk.c to produce a less verbose panic() dump,
	  and also to increase the number of lines of stack trace to 10.
	* Various spelling corrections to printk() messages.

===============================================================================

Thu Jan 02 16:12:36 EET 2003	Harry Kalogirou <harkal@gmx.net>

        * Better select hash.
	* This is ELKS 0.1.2

-------------------------------------------------------------------------------

Fri Nov 01 01:52:44 EET 2002	Harry Kalogirou <harkal@gmx.net>

        * Changed the position of ds in the _registers struct.
        * Added support for loading the new ELKS executable format. This is 

elks/fs/msdos/dir.c
/*
 *  linux/fs/msdos/dir.c
 *
 *  Written 1992 by Werner Almesberger
 *
 *  MS-DOS directory handling functions
 */
/* ported from linx-0.97 by zouys, Oct 21st, 2010 */
/* neated from linx-2.0.34 fangzs, Nov 23rd, 2010 */

#include <linuxmt/fs.h>
#include <linuxmt/msdos_fs.h>
#include <linuxmt/errno.h>
#include <linuxmt/stat.h>
#include <linuxmt/mm.h>
#include <linuxmt/string.h>
#include <linuxmt/debug.h>

#include <arch/segment.h>

static size_t msdos_dir_read(struct inode *dir, struct file *filp, char *buf, size_t count)
{
    return -EISDIR;
}

static int msdos_readdir(struct inode *dir, struct file *filp,
			 char *dirbuf, filldir_t filldir);

static struct file_operations msdos_dir_operations = {
	NULL,			/* lseek - default */
	msdos_dir_read,		/* read */
	NULL,			/* write - bad */
	msdos_readdir,		/* readdir */
	NULL,			/* select - default */
	NULL,			/* ioctl - default */
	NULL,			/* no special open code */
	NULL			/* no special release code */
};

/*
 * directories can handle most operations...
 */
struct inode_operations msdos_dir_inode_operations = {
	&msdos_dir_operations,	/* default directory file-ops */
	msdos_create,		/* create */
	msdos_lookup,		/* lookup */
	NULL,			/* link */
	msdos_unlink,		/* unlink */
	NULL,			/* symlink */
	msdos_mkdir,		/* mkdir */
	msdos_rmdir,		/* rmdir */
	NULL,			/* mknod */
	NULL,			/* readlink */
	NULL,			/* follow_link */
	NULL,			/* getblk */
	NULL			/* truncate */
};

static int FATPROC
unicode_to_ascii(char *ascii, unsigned char *uni)
{
	char *op = ascii;

	while(*uni && *uni!=0xff){
		*op++ = *uni;
		uni += 2;
	}
	*op = 0;
	return (op - ascii);
}

/*
 * Read next directory entry into name/namelen
 *	returns 1 on short filename
 *	returns 2 on long filename
 */
int FATPROC msdos_get_entry_long(
	struct inode *dir,		/* scan this directory*/
	off_t *pos,				/* starting at this offset*/
	struct buffer_head **bh,/* using this buffer*/
	char *name,				/* return entry name*/
	int *namelen,			/* return entry length*/
	off_t *dirpos,			/* return entry directory position*/
	ino_t *ino)				/* return entry inode*/
{
	struct msdos_dir_entry *de;
	off_t oldpos = *pos;	/* location of next directory entry start*/
	int is_long;
	unsigned char alias_checksum = 0;
	ASYNCIO_REENTRANT unsigned short unicodename[26+1];	/* Limited to two long entries */

	if ((int)*pos & (sizeof(struct msdos_dir_entry) - 1)) return -ENOENT;
	is_long = 0;
	*ino = msdos_get_entry(dir,pos,bh,&de);
	debug_fat("get_entry_long block %lu\n", EBH(*bh)->b_blocknr);
	while (*ino != (ino_t)-1L) {
		if (de->name[0] == 0) {		/* empty  entry and stop reading*/
			break;
		} else if ((unsigned char)de->name[0] == DELETED_FLAG) {	/* empty entry*/
			is_long = 0;
			oldpos = *pos;
		} else if (de->attr ==  ATTR_EXT) {		/* long filename entry*/
			int slot = 0;
			register struct msdos_dir_slot *ds = (struct msdos_dir_slot *) de;

			if (ds->id & LAST_LONG_ENTRY) {		/* last entry is first*/
				slot = ds->id & ~LAST_LONG_ENTRY;
				is_long = 1;
				alias_checksum = ds->alias_checksum;
			}

			unicodename[26] = 0;		        /* nul terminate unicode buffer*/
			while (slot > 0) {					/* read entries until slot 0*/
				if (ds->attr != ATTR_EXT ||
					(ds->id & ~LAST_LONG_ENTRY) != slot ||
					 ds->alias_checksum != alias_checksum) {
					is_long = 0;
					break;
				}
				/*
				 * Only save last two slot contents as limited to 14 char filenames
				 */
				slot--;
				if (slot < 2) {
					int offset = slot * 13;		/* 13 chars/entry*/
					memcpy(&unicodename[offset], ds->name0_4, 10);
					offset += 5;
					memcpy(&unicodename[offset], ds->name5_10, 12);
					offset += 6;
					memcpy(&unicodename[offset], ds->name11_12, 4);
					offset += 2;
					if (ds->id & LAST_LONG_ENTRY)
						unicodename[offset] = 0;
				}
				if (slot > 0) {
					*ino = msdos_get_entry(dir,pos,bh,&de);
					if (*ino == (ino_t)-1L) {
						is_long = 0;
						break;
					}
					ds = (struct msdos_dir_slot *)de;
				}
			}
		} else if (de->name[0] && !(de->attr & ATTR_VOLUME)) {	/* short filename entry*/
			int i,i2,last;
			int long_len = 0;
			unsigned char c;
			ASYNCIO_REENTRANT char longname[14];

			if (is_long) {
				unsigned char sum;
#pragma GCC diagnostic ignored "-Waggressive-loop-optimizations"
				for (i = sum = 0; i < MSDOS_NAME; i++) {
					sum = ((sum&1)<<7) | ((sum&0xfe)>>1);
					sum += de->name[i];
				}

				if (sum != alias_checksum)
					is_long = 0;

				unicodename[14] = 0;		/* truncate name to 14 characters*/
				long_len = unicode_to_ascii(longname, (unsigned char *)unicodename);
				/* FIXME should handle:
				 *	leading and trailing spaces ignored
				 *	trailing periods ignored
				 */
			}

			for (i = last = 0; i < 8; i++) {
				if (!(c = de->name[i])) break;
				if (c == 0x05) c = 0xE5;	/* 05 -> E5 as same as deleted entry*/
				if (c != ' ') last = i+1;
				name[i] = tolower(c);
			}
			i = last;
			name[i] = '.';
			i++;
			for (i2 = 0; i2 < 3; i2++) {
				if (!(c = de->ext[i2])) break;
				name[i] = tolower(c);
				i++;
				if (c != ' ') last = i;
			}
			if ((i = last) != 0) {
				if (!strcmp(de->name,MSDOS_DOT))
					*ino = dir->i_ino;
				else if (!strcmp(de->name,MSDOS_DOTDOT))
					*ino = msdos_parent_ino(dir,0);

				debug_fat("dir: '%s' attr %x size %lx\n", de->name, de->attr, de->size);
				if (is_long) {
					*namelen = long_len;
					*dirpos = oldpos;
					memcpy(name, longname, long_len);
					return 2;
				} else {
					*namelen = i;
					*dirpos = oldpos;
					return 1;
				}
			}
			is_long = 0;
		} else {
			is_long = 0;
			oldpos = *pos;
		}
		*ino = msdos_get_entry(dir,pos,bh,&de);

	} /* while (*ino != -1) */

	return 0;
}

/* Read a complete directory entry for a (specified) file,
 * submit a message to the callback function,
 * return a value of 0 or an error code.
 */
static int msdos_readdir(struct inode *dir, struct file *filp, char *dirbuf,
	filldir_t filldir)
{
	struct buffer_head *bh = NULL;
	ino_t ino;
	off_t dirpos;
	int res, namelen;
	ASYNCIO_REENTRANT char name[14];

	if (!dir || !S_ISDIR(dir->i_mode)) return -EBADF;
	if (dir->i_ino == MSDOS_ROOT_INO) {
		/* Fake . and .. for the root directory*/
		if ((int)filp->f_pos < 2) {
			/* Tricky: returns "." or ".." depending on namelen*/
			size_t namlen = (size_t)++filp->f_pos;
			return filldir(dirbuf, (char *)"..", namlen, filp->f_pos, (ino_t)MSDOS_ROOT_INO);
		}
		if ((int)filp->f_pos == 2)	/* reset to 0 after '..' for real directory offset*/
			filp->f_pos = 0;
	}

	res = msdos_get_entry_long(dir, &filp->f_pos, &bh, name, &namelen, &dirpos, &ino);
	if (res > 0)
		filldir(dirbuf, name, namelen, dirpos, ino);
	if (bh)
		unmap_brelse(bh);
	if (res < 0)
		return res;
	return 0;
}
